cpp/src/arrow/dataset/scanner.h

```diff
@@ -138,47 +179,147 @@ ARROW_DS_EXPORT Result<ScanTaskIterator> ScanTaskIteratorFromRecordBatch(
     std::vector<std::shared_ptr<RecordBatch>> batches,
     std::shared_ptr<ScanOptions> options);
 
-/// \brief Scanner is a materialized scan operation with context and options
-/// bound. A scanner is the class that glues ScanTask, Fragment,
-/// and Dataset. In python pseudo code, it performs the following:
+template <typename T>
+struct Enumerated {
+  T value;
+  int index;
+  bool last;
+};
+
+/// \brief Combines a record batch with the fragment that the record batch originated
+/// from
+///
+/// Knowing the source fragment can be useful for debugging & understanding loaded data
```

Comment:
Indeed.

Reply:
I'll backport this into #9589 since they need to be synced up at some point.

Reply:
Sorry about that.  If you want to land #9589 first I can rebase this on top of that.  I'm pretty sure the two are fairly compatible although there is some work to adjust to the new structs.

Reply:
I've already backported it so whichever one gets reviewed first should get merged first :) (though I'm about to go back and remove `operator==` as per your point below.)